File: DPCodes/55-jumpGame.py
from functools import cache
class Solution:
    def canJump(self, nums):# List[int]) -> bool:
        n=len(nums)
        @cache
        def dp(i):
            if i>=n-1:
                return True
            if i+nums[i]>=n-1:
                return True
            for s in range(1,nums[i]+1):
                if dp(s+i):
                    return True
            return False
        return dp(0)
    
nums=[3,2,1,1,4]
print(Solution().canJump(nums))

File: DPCodes/55-jumpGame.cpp
#include<iostream>
#include<vector>
#include<cmath>

using std::cout;
using VEC=std::vector<int>;


bool foo(const VEC &nums,int i)
{
    int n=nums.size();
    if(i>=n-1)
        return true;
    if(i+nums[i]>=n-1)
        return true;
    
    for(int s=1;s<nums[i]+1;++s)  //the end value should be nums[i]+1
    {
        if(foo(nums,s+i))
            return true;
    }
    return false;
}

int canJump(const VEC& nums)
{
    return foo(nums,0);
}

int main()
{
    VEC nums{3,2,1,0,4};
    cout<<canJump(nums);
}